url5a01f2c1d9_core9_core_tgz-pJ8-io_core9_core_boot_BootstrapFrameworkJ8/src/net/xeoh/plugins/base/impl/classpath/cache/JARCache.java

```diff
@@ -0,0 +1,298 @@
+			final Object rval = ois.readObject(); // Loading caches can take
+												  // very long
+			                                      // (>250ms) for many entries
+
+			ois.close();
+
+			if (rval != null) {
+				this.cacheMap = (Map<String, JARInformation>) rval;
+			}
+		} catch (EOFException e) {
+			//
+		} catch (final FileNotFoundException e) {
+			// If file was not found, no problem ...
+			// e.printStackTrace();
+		} catch (final IOException e) {
+			e.printStackTrace();
+		} catch (final ClassNotFoundException e) {
+			this.logger
+			        .warning("Your JSPF cache is outdated, please delete it. It will be regenerated with the next run. The next exception reflects this, so don't be afraid.");
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * @param uri
+	 * @return .
+	 */
+	public JARInformation getJARInformationFor(final URI uri) {
+		// Just return a dummy when we are disabled
+		if (!this.cacheEnabled) {
+			return new JARInformation();
+		}
+
+		final File file = new File(uri);
+		final String hash = (this.weakMode) ? generateWeakHash(file) : generateStrongHash(file);
+
+		if (hash == null) {
+			this.logger.warning("Error generating hash. Caching won't work.");
+			return new JARInformation();
+		}
+
+		// Create information if not already there.
+		if (!this.cacheMap.containsKey(hash)) {
+			this.cacheMap.put(hash, new JARInformation());
+		}
+
+		final JARInformation jarInformation = this.cacheMap.get(hash);
+		jarInformation.lastAccess = System.currentTimeMillis();
+		jarInformation.usageCount++;
+		return jarInformation;
+	}
+
+	/**
+	 * Saves the cache
+	 */
+	public void saveCache() {
+
+		if (!this.cacheEnabled)
+			return;
+		String cacheFile = this.cachePath;
+		// String = this.configuration.getConfiguration(getClass(),
+		// "cacheFile");
+		if (cacheFile == null) {
+			cacheFile = DEFAULT_CACHE_FILE;
+		}
+
+		try {
+			// lock the lockfile
+			final FileOutputStream fos = new FileOutputStream(cacheFile);
+			@SuppressWarnings("resource")
+			final ObjectOutputStream oos = new ObjectOutputStream(fos);
+
+			oos.writeObject(this.cacheMap);
+		} catch (final FileNotFoundException e) {
+			e.printStackTrace();
+		} catch (final IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	/**
+	 * Weak hash mode.
+	 * 
+	 * @param file
+	 * @return
+	 */
+	private String generateWeakHash(File file) {
+		final String name = file.getName();
+		final long length = file.length();
+
+		final String hash = "weak:" + name + "@" + length;
+		return hash;
+	}
+
+	/**
+	 * Generate a hash for the given element.
+	 * 
+	 * @param element
+	 * 
+	 * @return .
+	 */
+	@SuppressWarnings("boxing")
+	private String generateStrongHash(final File element) {
+		// Try to generate hash
+
+		boolean created = false;
+
+		try {
+			this.logger.fine("Processing JAR " + element);
+
+			final MessageDigest digest = java.security.MessageDigest.getInstance("MD5");
+			final FileInputStream fis = new FileInputStream(element);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at lines 229, 235, 238. The resource is closed at line 251. There are other execution paths that do not contain closure statements, for example, when FileInputStream.read throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).